### PR TITLE
health: move to mypy

### PR DESCRIPTION
**Pull request by WilliamBergamin**
[MyPy](https://mypy.readthedocs.io/en/latest/index.html) was found to execute much fast the [PyType](https://google.github.io/pytype/) in this project, these changes adapt the project to use MyPy instead of PyType

| pytype           | mypy |
| -----------  | ------ |
| Screenshot 2024-08-15 at 5 45 47 PM | Screenshot 2024-08-15 at 5 46 03 PM | 
| 11m27s | 43s |

From my short experience with mypy:
- Good documentation 📜 
- Good error surfacing
- Easy to configure
- It offers a language server, this helps with development since it integrate with IDEs

Concern: This tool can be strict if we let it, I think it will lead to "safer", more resilient, code. We should not let it get in the way of writing clear and efficient code.

Note: removing all the `type: ignore` comments from the project may cause breaking changes or have a ripple effect through the entire project

Let me know If we want more tests to cover these changes 🤔 

---
### Category

* [ ] `slack_bolt.App` and/or its core components
* [ ] `slack_bolt.async_app.AsyncApp` and/or its core components
* [ ] Adapters in `slack_bolt.adapter`
* [ ] Document pages under `/docs`
* [x] Others

## Requirements

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've run `./scripts/install_all_and_run_tests.sh` after making the changes.

### CODE REVIEW

File: slack_bolt/middleware/authorization/async_multi_teams_authorization.py
Comment (WilliamBergamin) [inline slack_bolt/middleware/authorization/async_multi_teams_authorization.py:99]: I don't think this should change the current behavior of the code base and it should make it more resilient, but asserting callable is another option
 ```suggestion
 if req.context.response_url is not None:
         assert callable(req.context.respond), "req.context.respond must be callable"
```

File: slack_bolt/oauth/async_oauth_flow.py
Comment (WilliamBergamin) [inline slack_bolt/oauth/async_oauth_flow.py:61]: `self.settings.logger` should never be `None`

**Comment by filmaj** [comment]
My goodness from 11 minutes to 40 seconds 🤯

File: slack_bolt/app/app.py
Comment (filmaj) [reply slack_bolt/app/app.py:112]: Seems odd to have an optional boolean, unless the None state has a special meaning and so `installation_store_bot_only` is meant to have one of three states?

**Review by filmaj — COMMENTED** [review]
Nicely done! The speed boosts are significant so it seems like a good change, but I will defer to Kaz's opinion on this - so will not put a checkmark on it 😄 sorry!

**Review by seratch — CHANGES_REQUESTED** [review]
@WilliamBergamin Great work! 

The reasons why I chose pytype over mypy during the initial development of bolt-python a few years ago are:
- I compared mypy, pyright, and pytype; for this project, pytype was the most intelligent. It detected many type hint issues by analyzing method calls within bolt-python and test code. Slow execution was already a concern, but running it on CI builds was acceptable. Pyright looked good too, but it was not as helpful for this project in terms of the number of error/mistake detections.
- Back then, mypy displayed way more unhelpful errors than other tools. During active development, I didn't have time to handle that. Also, pytype was significantly smarter for many patterns then.

Pytype was incredibly helpful, especially for the initial development, but I notice that it does not detect many issues anymore because this project is very stable and there are fewer type hint errors now. For this stage, I am open to prioritizing execution speed over the linter's intelligence.

File: slack_bolt/app/app.py
Comment (WilliamBergamin) [reply slack_bolt/app/app.py:420]: Great suggestion, I've added a BoltError exception, but I'm unsure of its message

```py
        else:
            raise BoltError("OAuthFlow or Authorize must be configured to make a Bolt app")
```

Let me know if this is sufficient 🤔

File: slack_bolt/listener/async_listener_error_handler.py
Comment (WilliamBergamin) [reply slack_bolt/listener/async_listener_error_handler.py:51]: This change was intentional, `mypy` believes that `response` can be None in this situation, but I can revert this

File: slack_bolt/middleware/authorization/async_multi_teams_authorization.py
Comment (WilliamBergamin) [reply slack_bolt/middleware/authorization/async_multi_teams_authorization.py:99]: Sounds good 💯 I will remove the `assert` changes and add `type: ignore` statements when needed

File: slack_bolt/logger/messages.py
Comment (WilliamBergamin) [reply slack_bolt/logger/messages.py:207]: I agree, it appears the following satisfies mypy

```py
req.body["actions"][0].get("action_id")
```

File: slack_bolt/logger/messages.py
Comment (WilliamBergamin) [reply slack_bolt/logger/messages.py:221]: Same here also it seems the following satisfies mypy

```py
constraints = '"' + req.body["action_id"] + '"'
```

**Comment by WilliamBergamin** [comment]
@seratch thank you for sharing the context behind originally choosing pytype over mypy, now that I've used both libraries I see how pytype is more intelligent and covers more patterns  then mypy, I assume this gap was larger in the past. I see how useful this can be in the early stages this project 🙇 

Simultaneously mypy did catch a few edge cases that I was not aware of and may be worth addressing in the future. For [example](https://github.com/slackapi/bolt-python/blob/40f6d1e46cdf8f958ad682c9ae0aeb73d27ce616/slack_bolt/context/context.py#L37) the `context.client` is `optional` this has ripple effects in the project since it regularly assumes the client value is not `None`

File: slack_bolt/app/app.py
Comment (seratch) [reply slack_bolt/app/app.py:420]: The error message looks good to me. For consistency, move the string to `slack_bolt.logger.messages` like other parts in this class do!

File: slack_bolt/app/app.py
Comment (seratch) [inline slack_bolt/app/app.py:430]: move the string to slack_bolt.logger.messages

File: slack_bolt/app/async_app.py
Comment (seratch) [inline slack_bolt/app/async_app.py:429]: same as above

**Review by seratch — APPROVED** [review]
Left one last comment about the error message string. After resolving it, you can merge this PR tomorrow in your timezone! Thanks for making these changes!

**Comment by seratch** [comment]
@WilliamBergamin Can you merge https://github.com/slackapi/bolt-python/pull/1132 first if it looks good to you? The changes by that PR may cause conflicts with this PR. If that's the case, please resolve it on this PR side 🙏

**Pull request merged**
